### PR TITLE
Test framework

### PR DESCRIPTION
**Pull request by junmendoza**
### Declarations
See the design here:
https://docs.google.com/document/d/1pG6DXp3iFWbavTElm1KZtWRRiBzkFXVl_ZDLzEnacHw/edit

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] The level of testing this PR includes is appropriate

### Reviewers

@lukechurch 

### FYIs

This is using JSON.net to parse the JSON formatted verification string.
Json.NET is open source under the MIT license and is free for commercial use.
http://www.newtonsoft.com/json

@RodRecker

### CODE REVIEW

File: test/Engine/ProtoTest/Associative/SimpleVerification.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            thisTest.RunAndVerify(
+                code,
+                TestFrameWork.BuildVerifyPair("a", 1),
+                TestFrameWork.BuildVerifyPair("b", 2)
+                );
+        }
+
+
+        [Test]
+        public void TestFunctionCall01()
```
Comment (ke-yu) [inline test/Engine/ProtoTest/Associative/SimpleVerification.cs]: How about this? 
```thisTest.RunAndVerify(
    @"a = 1; b = {1,2,3}";
    @"{"a":1, "b":[1,2,3]}")
```?

:-)

Comment (junmendoza) [inline test/Engine/ProtoTest/Associative/SimpleVerification.cs]: Yea we can go directly to the string format. Ill define the string formatting in the doc itself:
https://docs.google.com/document/d/1pG6DXp3iFWbavTElm1KZtWRRiBzkFXVl_ZDLzEnacHw/edit

File: src/Engine/ProtoCore/BuildStatus.cs [hunk 2]
```diff
@@ -26,6 +26,12 @@ public BuildHaltException(string message)
 
     namespace BuildData
     {
+        public enum ErrorID
+        {
+            kDefault,
```
Comment (lukechurch) [inline src/Engine/ProtoCore/BuildStatus.cs]: k_Nonsense.

Not on new enums, please.

File: test/Engine/ProtoTest/Associative/SimpleVerification.cs (hunk 1 above)
Comment (lukechurch) [inline test/Engine/ProtoTest/Associative/SimpleVerification.cs]: Could we use something more like JSON?

That way we get better portability, pretty printers for free, parsers etc.

The only think that we wouldn't get that I can see is indexed lookup.

Comment (junmendoza) [inline test/Engine/ProtoTest/Associative/SimpleVerification.cs]: Will explore this JSON

Comment (junmendoza) [inline test/Engine/ProtoTest/Associative/SimpleVerification.cs]: The testcases using JSON format is here:

VerificationFormatJSON.cs

The parser is currently just a stub

**Comment by lukechurch** [comment]
Any commentary on why the build is broken with this?

**Comment by lukechurch** [comment]
Need to modify the readme to include a notification of the 3rd party component, and cc @RodRecker to let him know

File: src/Engine/ProtoCore/BuildStatus.cs (hunk 2 above)
Comment (lukechurch) [inline src/Engine/ProtoCore/BuildStatus.cs]: ping. Please can we change this to normal C# format?

File: src/Engine/ProtoCore/BuildStatus.cs [hunk 3]
```diff
@@ -26,6 +26,12 @@ public BuildHaltException(string message)
 
     namespace BuildData
     {
+        public enum ErrorID
```
Comment (lukechurch) [inline src/Engine/ProtoCore/BuildStatus.cs]: This isn't really an ID? It's more a type of error?

File: test/Engine/ProtoTest/Associative/VerificationFormatJSON.cs [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            string verification =
+@" 
+    {
        ""pi"" : 3.14,
        ""e"" : 2.71828,
+    }
+";
+            thisTest.RunAndVerify(code, verification);
+
+        }
```
Comment (lukechurch) [inline test/Engine/ProtoTest/Associative/VerificationFormatJSON.cs]: Lets use  a 1 line format where it makes sense { "'a"" : 1.0 }

**Comment by lukechurch** [comment]
@junmendoza OK, this seems to be heading in right direction with the verification format. Thanks!

Can you push forwards a bit on the parser implementation and let me know when you get there?

File: src/Engine/ProtoCore/BuildStatus.cs (hunk 2 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/BuildStatus.cs]: kDone kCheckedIn :(

File: src/Engine/ProtoCore/BuildStatus.cs (hunk 3 above)
Comment (junmendoza) [inline src/Engine/ProtoCore/BuildStatus.cs]: Yes it is the error type. However it just follows the current convention in BuildStatus
such as public enum WarningID.

Its easy to modify but theres gonna be hundreds of changes, so i think once you LGTM this, i can then modify the constants so it doesnt litter the review this early?

File: test/Engine/ProtoTest/Associative/VerificationFormatJSON.cs (hunk 4 above)
Comment (junmendoza) [inline test/Engine/ProtoTest/Associative/VerificationFormatJSON.cs]: done

**Comment by junmendoza** [comment]
@lukechurch The parser is implemented in JsonDecoder. It now works when verifying our types.

**Comment by lukechurch** [comment]
@junmendoza why is the build broken with this PR? I'd really like to understand that before we merge.

**Comment by junmendoza** [comment]
Build is fixed

File: test/Engine/ProtoTest/Associative/VerificationFormat.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                "i:2|a:{1,2,5}",
+                TestFrameWork.VerificationFormat.Custom
+                );
+        }
+
+        [Test]
+        [Ignore] // Custom verification format for arrays not implemented
+        public void TestNestedArrayAssignment01()
+        {
+            string code =
```
Comment (lukechurch) [inline test/Engine/ProtoTest/Associative/VerificationFormat.cs]: Lets just use JSON format and get rid of these

File: src/Engine/ProtoCore/BuildStatus.cs (hunk 3 above)
Comment (lukechurch) [inline src/Engine/ProtoCore/BuildStatus.cs]: Lets accept the inconsistency for now

File: test/Engine/ProtoTestFx/ProtoTestFx.csproj [hunk 6]
```diff
@@ -79,6 +80,10 @@
     <AssemblyOriginatorKeyFile>ProtoTestFx.snk</AssemblyOriginatorKeyFile>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="Newtonsoft.Json, Version=4.5.0.0, Culture=neutral, processorArchitecture=MSIL">
```
Comment (lukechurch) [inline test/Engine/ProtoTestFx/ProtoTestFx.csproj]: Lets use the existing check in of this.

**Comment by lukechurch** [comment]
Given changes discussed, LGTM, lets get this in and start porting some of the tests

**Pull request merged**